Zend/zend_execute.h

```diff
@@ -179,6 +179,37 @@ static zend_always_inline zval* zend_assign_to_variable(zval *variable_ptr, zval
 	return variable_ptr;
 }
 
+static zend_always_inline zval* zend_assign_to_variable_twice(zval *variable_ptr, zval *second_variable_ptr, zval *value, uint8_t value_type, bool strict)
+{
+	do {
+		if (UNEXPECTED(Z_REFCOUNTED_P(variable_ptr))) {
+			zend_refcounted *garbage;
+
+			if (Z_ISREF_P(variable_ptr)) {
+				if (UNEXPECTED(ZEND_REF_HAS_TYPE_SOURCES(Z_REF_P(variable_ptr)))) {
+					value = zend_assign_to_typed_ref(variable_ptr, value, value_type, strict);
+					ZVAL_COPY(second_variable_ptr, value);
+					return value;
+				}
+
+				variable_ptr = Z_REFVAL_P(variable_ptr);
+				if (EXPECTED(!Z_REFCOUNTED_P(variable_ptr))) {
+					break;
+				}
+			}
+			garbage = Z_COUNTED_P(variable_ptr);
+			zend_copy_to_variable(variable_ptr, value, value_type);
+			GC_DTOR_NO_REF(garbage);
+			ZVAL_COPY(second_variable_ptr, variable_ptr);
```

Comment:
Isn't the whole point of this to assign to `second_variable_ptr` before calling `GC_DTOR_NO_REF(garbage)`, as that may reassign the variable, freeing `variable_ptr` before being assigned again? I think that's the difference in the tests you're seeing, this should assign `value` directly, preferably, and before `GC_DTOR_NO_REF` (`variable_ptr` and `value` should be the same at that point).